### PR TITLE
More higher order variable checks and support for transfers

### PR DESCRIPTION
**Pull request by GiudGiud**
This PR will have to wait for Positions to go in, it conflicts
EDIT: all good now

### CODE REVIEW

File: framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C [hunk 1]
```diff
@@ -56,6 +63,84 @@ MultiAppGeneralFieldNearestNodeTransfer::MultiAppGeneralFieldNearestNodeTransfer
...
+      if (from_var.isNodal())
+        _source_is_nodes[var_index] = true;
+      // for everyone else, we know they are continuous at centroids
+      else
+        _source_is_nodes[var_index] = false;
+    }
+
+    // Some variables can be sampled directly at their 0 dofs
+    if ((_source_is_nodes[var_index] && fe_type.family == LAGRANGE) ||
+        (!_source_is_nodes[var_index] && fe_type.family == MONOMIAL && fe_type.family == CONSTANT))
```
Comment (GiudGiud) [inline framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C:107]: @roystgnr 

I m guessing I m too restrictive here. But there's also 20 something element types and 10 families 
I ll see if I can find a good output to see this

Comment (GiudGiud) [reply framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C:107]: ok I think I just need to take a deep dive here
https://mooseframework.inl.gov/docs/doxygen/libmesh/generic__projector_8h_source.html

Comment (roystgnr) [reply framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C:107]: Even `LAGRANGE` is tricky - if you have a superparametric element (1st order FE on 2nd order Elem), which is super popular (e.g. the pressure variable in Taylor-Hood), you can't just evaluate component 0 at non-vertex nodes.

Any C0-continuous (including C1-continuous) scalar element in libMesh will take a value of nodal-DoF-0 on vertices.  Any discontinuous `CONSTANT` element will take a value of elemental-DoF-0 everywhere.  Isoparametric `LAGRANGE` will take a value of nodal-DoF-0 on non-vertex nodes.  Vector-valued elements do something similar component-wise but with offsets for different components.  And I think that may be an exhaustive list.

Comment (GiudGiud) [reply framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C:107]: oh that first item is going to be tricky

Comment (GiudGiud) [reply framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C]: ok was not too bad.

So I m correct in assuming that at all orders, the value at the centroid of a MONOMIAL is the same as the dof 0 value (in my mind, all the shape functions except the constant one are 0 there)

File: framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C [hunk 2]
```diff
@@ -143,13 +251,20 @@ MultiAppGeneralFieldNearestNodeTransfer::buildKDTrees(const unsigned int var_ind
           if (!_from_boundaries.empty() && !onBoundaries(_from_boundaries, from_mesh, elem))
             continue;
 
+          const auto centroid = elem->vertex_average();
```
Comment (lindsayad) [inline framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C:254]:
```suggestion
          const auto vertex_average = elem->vertex_average();
```

File: framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C [hunk 3]
```diff
@@ -143,13 +251,20 @@ MultiAppGeneralFieldNearestNodeTransfer::buildKDTrees(const unsigned int var_ind
...
+
           if (_nearest_positions_obj &&
-              !closestToPosition(i_source, elem->vertex_average() + _from_positions[i_from]))
+              !closestToPosition(i_source, centroid + _from_positions[i_from]))
```
Comment (lindsayad) [inline framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C:257]:
```suggestion
              !closestToPosition(i_source, vertex_average + _from_positions[i_from]))
```

File: framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C [hunk 4]
```diff
@@ -143,13 +251,20 @@ MultiAppGeneralFieldNearestNodeTransfer::buildKDTrees(const unsigned int var_ind
...
             continue;
 
-          const auto dof = elem->dof_number(from_sys.number(), from_var_num, 0);
-          _local_points[i_source].push_back(elem->vertex_average() + _from_positions[i_from]);
-          _local_values[i_source].push_back((*from_sys.solution)(dof));
+          _local_points[i_source].push_back(centroid + _from_positions[i_from]);
```
Comment (lindsayad) [inline framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C:260]:
```suggestion
          _local_points[i_source].push_back(vertex_average + _from_positions[i_from]);
```

File: framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C [hunk 5]
```diff
@@ -143,13 +251,20 @@ MultiAppGeneralFieldNearestNodeTransfer::buildKDTrees(const unsigned int var_ind
...
+          if (_use_zero_dof_for_value[var_index])
+          {
+            auto dof = elem->dof_number(from_sys.number(), from_var_num, 0);
+            _local_values[i_source].push_back((*from_sys.solution)(dof));
+          }
+          else
+            _local_values[i_source].push_back(from_sys.point_value(from_var_num, centroid, elem));
```
Comment (lindsayad) [inline framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C:267]:
```suggestion
            _local_values[i_source].push_back(from_sys.point_value(from_var_num, vertex_average, elem));
```

File: framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C [hunk 6]
```diff
@@ -124,11 +220,23 @@ MultiAppGeneralFieldNearestNodeTransfer::buildKDTrees(const unsigned int var_ind
               !closestToPosition(i_source, *node + _from_positions[i_from]))
             continue;
 
-          const auto dof = node->dof_number(from_sys.number(), from_var_num, 0);
           _local_points[i_source].push_back(*node + _from_positions[i_from]);
-          _local_values[i_source].push_back((*from_sys.solution)(dof));
+          if (_use_zero_dof_for_value[var_index])
+          {
+            auto dof = node->dof_number(from_sys.number(), from_var_num, 0);
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C:226]: General Field does not support vector variables

File: framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C [hunk 7]
```diff
@@ -124,11 +220,23 @@ MultiAppGeneralFieldNearestNodeTransfer::buildKDTrees(const unsigned int var_ind
...
+            _local_values[i_source].push_back((*from_sys.solution)(dof));
+          }
+          else
+          {
+            flagInvalidSolution(
+                "Nearest-location is not implemented for this source variable type on "
+                "this mesh. Returning value at dof 0");
+            auto dof = node->dof_number(from_sys.number(), from_var_num, 0);
+            _local_values[i_source].push_back((*from_sys.solution)(dof));
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C]: done

**Review by lindsayad — DISMISSED** [review]

File: framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C (hunk 1 above)
Comment (roystgnr) [reply framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C]: No, I'm afraid.  On elements with linear mappings, this should be true, but even an element as simple as a trapezoid will be skewed differently.  I think with a bilinear quad you'll see the dof 0 value at the `vertex_average()` point, but not at the `true_centroid()` point in general.

Comment (roystgnr) [reply framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C]: Oh, wait, even the above isn't pessimistic enough!

Comment (roystgnr) [reply framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C]: So, `MONOMIAL` elements will take their DoF-0 value at `Point(0,0,0)` in master space.  For Edge/Quad/Hex elements, our master elements are just tensor products of the interval `[-1,1]`, and `(0,0,0)` is the centroid of that, and if they have linear (affine) mappings to physical space then the centroid of the mapped element is the mapping of the centroid, and it all works out.  If the mapping isn't linear, then the vertex average isn't the centroid, as I said.

But if the master element isn't a tensor product element?  We don't even use the same convention!  A triangle, for example, has vertices `(0,0)`, `(1,0)`, and `(0,1)` in master space.  Not only does DoF 0 not end up evaluating on the centroid, it barely ends up evaluating on the element at all!  The physical element's vertex 0 ends up being where DoF 0's value matches the physical value.

Comment (roystgnr) [reply framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C]: Prisms see master (0,0,0) in the middle of one edge, Pyramids see it in the middle of the square face ... I think the safest thing is to just do the weighted shape function sum for `p>0`; by the time you've tested `has_affine_map()` and `type()` of an `Elem` to see if that's safe to shortcut, you've already done half as much work anyway.

Comment (GiudGiud) [reply framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C]: removed monomials from the allow-list

**Review by lindsayad — APPROVED** [review]
Does @roystgnr approve?

**Review by roystgnr — APPROVED** [review]

**Comment by GiudGiud** [comment]
Thanks for all the reviews!
I'll try to synthesize this knowledge you distilled on higher order in a doc page some time

**Pull request merged**